internal/self_metrics/metrics.go

```diff
@@ -0,0 +1,135 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package self_metrics
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"time"
+
+	mexporter "github.com/GoogleCloudPlatform/opentelemetry-operations-go/exporter/metric"
+	"github.com/GoogleCloudPlatform/ops-agent/confgenerator"
+	"go.opentelemetry.io/contrib/detectors/gcp"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/metric/instrument"
+	"go.opentelemetry.io/otel/sdk/metric/controller/basic"
+	"go.opentelemetry.io/otel/sdk/metric/sdkapi"
+	"go.opentelemetry.io/otel/sdk/resource"
+)
+
+var (
+	agentMetricsPrefixFormat = "agent.googleapis.com/%s"
+	formatter                = func(d *sdkapi.Descriptor) string {
+		return fmt.Sprintf(agentMetricsPrefixFormat, d.Name())
+	}
+)
+
+type enabledReceivers struct {
+	metric map[string]int
+	log    map[string]int
+}
+
+func GetEnabledReceivers(uc *confgenerator.UnifiedConfig) (enabledReceivers, error) {
+	eR := enabledReceivers{
+		metric: make(map[string]int),
+		log:    make(map[string]int),
+	}
+
+	// Logging Pipelines
+	for _, p := range uc.Logging.Service.Pipelines {
+		for _, rID := range p.ReceiverIDs {
+			rType := uc.Logging.Receivers[rID].Type()
+			eR.log[rType] += 1
+		}
+	}
+
+	// Metrics Pipelines
+	for _, p := range uc.Metrics.Service.Pipelines {
+		for _, rID := range p.ReceiverIDs {
+			rType := uc.Metrics.Receivers[rID].Type()
+			eR.metric[rType] += 1
+		}
+	}
+
+	return eR, nil
+}
+
+func CollectOpsAgentSelfMetrics(uc *confgenerator.UnifiedConfig, death chan bool) error {
+	eR, err := GetEnabledReceivers(uc)
+	if err != nil {
+		return err
+	}
+	log.Println("Found Enabled Receivers : ", eR)
+
+	// Detect GCP credentials
+	ctx := context.Background()
+	res, err := resource.New(ctx,
+		resource.WithDetectors(gcp.NewDetector()),
+		resource.WithTelemetrySDK(),
+	)
+
+	resOpt := basic.WithResource(res)
+
+	// Exporter options
+	opts := []mexporter.Option{
+		mexporter.WithMetricDescriptorTypeFormatter(formatter),
+		mexporter.WithInterval(time.Minute),
+	}
+
+	// Create exporter pipeline
+	pusher, err := mexporter.InstallNewPipeline(opts, resOpt)
+	if err != nil {
+		return fmt.Errorf("Failed to establish pipeline: %v", err)
+	}
+	defer pusher.Stop(ctx)
+
+	// Start meter
+	meter := pusher.Meter("ops_agent/self_metrics")
+
+	// Collect GAUGE metric
+	gaugeObserver, err := meter.AsyncInt64().Gauge("agent/ops_agent/enabled_receivers")
```

Comment:
For our case i think we could just use SyncInt64()? 🤔 @dashpole to confirm

Reply:
I tried using `SyncInt64()`, but it doesn't have support for `GAUGE` metrics ([link](https://github.com/open-telemetry/opentelemetry-go/blob/18f4cb85ece82b12cb9bd9af02efe2a47bd8f76e/metric/instrument/syncint64/syncint64.go) ) but `AsyncInt64()` does have it ([link](https://github.com/open-telemetry/opentelemetry-go/blob/18f4cb85ece82b12cb9bd9af02efe2a47bd8f76e/metric/instrument/asyncint64/asyncint64.go))

Reply:
Ah, thats right.  AsyncInt64 is probably what you want here.  If you don't want to check something too often, you should be able to implement something inside the callback that returns the previous value if it hasn't been long enough.  But its probably better to start with fetching it each time the metric is sent to start...

Reply:
Ok! That makes sense. I think the way it is implemented right now, there is no further way to do more precalculations for the metric, it would stay this way (looping through every value and labels combination to call the Observe() method). Thanks for clarifying this.